StackOverflowException via PatternMatcher

The implementation of `org.gradle.api.internal.file.pattern.PatternMatcher#or` via `org.gradle.api.internal.file.pattern.PatternMatcherFactory#getPatternsMatcher` (and `and()`) create a new anonymous class that calls the or-ed pattern recursively. If the number of tests and patterns to compare against is large enough, this implementation can likely result in a `java.lang.StackOverflowException`.

In my case, it happened when Gradle (version 5.6) was checking the output of an annotation processor against the cached elements, even if the stack size was manually set to 2048k.

A way to easily fix the j.l.StackOverflowException is to tweak `PatternMatcher` by implementing an inner `Or` class (and `And` class) that collects the `patternMatcher`s to be "or-ed" and exercises the "many OR operation" on the patterns collected in a list.
Another way would have been to collect the patterns in [`org.gradle.api.internal.file.pattern.PatternMatcherFactory#getPatternsMatcher` and implement a custom `PatternMatcher` there](https://github.com/snazy/gradle/commit/158b31f03d89891da8db240e118b524057c98852), but the former approach feels more general.

### Expected Behavior
No j.l.StackOverflowException

### Current Behavior
```
org.gradle.api.tasks.TaskExecutionException: Execution failed for task ':xyz'.
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter$3.accept(ExecuteActionsTaskExecuter.java:166)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter$3.accept(ExecuteActionsTaskExecuter.java:163)
        at org.gradle.internal.Try$Failure.ifSuccessfulOrElse(Try.java:191)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.execute(ExecuteActionsTaskExecuter.java:156)
        at org.gradle.api.internal.tasks.execution.ValidatingTaskExecuter.execute(ValidatingTaskExecuter.java:62)
        at org.gradle.api.internal.tasks.execution.SkipEmptySourceFilesTaskExecuter.execute(SkipEmptySourceFilesTaskExecuter.java:108)
        at org.gradle.api.internal.tasks.execution.ResolveBeforeExecutionOutputsTaskExecuter.execute(ResolveBeforeExecutionOutputsTaskExecuter.java:67)
        at org.gradle.api.internal.tasks.execution.StartSnapshotTaskInputsBuildOperationTaskExecuter.execute(StartSnapshotTaskInputsBuildOperationTaskExecuter.java:62)
        at org.gradle.api.internal.tasks.execution.ResolveAfterPreviousExecutionStateTaskExecuter.execute(ResolveAfterPreviousExecutionStateTaskExecuter.java:46)
        at org.gradle.api.internal.tasks.execution.CleanupStaleOutputsExecuter.execute(CleanupStaleOutputsExecuter.java:94)
        at org.gradle.api.internal.tasks.execution.FinalizePropertiesTaskExecuter.execute(FinalizePropertiesTaskExecuter.java:46)
        at org.gradle.api.internal.tasks.execution.ResolveTaskExecutionModeExecuter.execute(ResolveTaskExecutionModeExecuter.java:95)
        at org.gradle.api.internal.tasks.execution.SkipTaskWithNoActionsExecuter.execute(SkipTaskWithNoActionsExecuter.java:57)
        at org.gradle.api.internal.tasks.execution.SkipOnlyIfTaskExecuter.execute(SkipOnlyIfTaskExecuter.java:56)
        at org.gradle.api.internal.tasks.execution.CatchExceptionTaskExecuter.execute(CatchExceptionTaskExecuter.java:36)
        at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter$1.executeTask(EventFiringTaskExecuter.java:77)
        at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter$1.call(EventFiringTaskExecuter.java:55)
        at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter$1.call(EventFiringTaskExecuter.java:52)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor$CallableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:416)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor$CallableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:406)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor$1.execute(DefaultBuildOperationExecutor.java:165)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:250)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:158)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.call(DefaultBuildOperationExecutor.java:102)
        at org.gradle.internal.operations.DelegatingBuildOperationExecutor.call(DelegatingBuildOperationExecutor.java:36)
        at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter.execute(EventFiringTaskExecuter.java:52)
        at org.gradle.execution.plan.LocalTaskNodeExecutor.execute(LocalTaskNodeExecutor.java:43)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$InvokeNodeExecutorsAction.execute(DefaultTaskExecutionGraph.java:355)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$InvokeNodeExecutorsAction.execute(DefaultTaskExecutionGraph.java:343)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$BuildOperationAwareExecutionAction.execute(DefaultTaskExecutionGraph.java:336)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$BuildOperationAwareExecutionAction.execute(DefaultTaskExecutionGraph.java:322)
        at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker$1.execute(DefaultPlanExecutor.java:134)
        at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker$1.execute(DefaultPlanExecutor.java:129)
        at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.execute(DefaultPlanExecutor.java:202)
        at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.executeNextNode(DefaultPlanExecutor.java:193)
        at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.run(DefaultPlanExecutor.java:129)
        at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:64)
        at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:48)
        at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:56)
Caused by: com.google.common.util.concurrent.ExecutionError: java.lang.StackOverflowError
        at com.google.common.cache.LocalCache$Segment.get(LocalCache.java:2047)
        at com.google.common.cache.LocalCache.get(LocalCache.java:3851)
        at com.google.common.cache.LocalCache$LocalManualCache.get(LocalCache.java:4713)
        at org.gradle.api.tasks.util.internal.CachingPatternSpecFactory$CachingSpec.isSatisfiedBy(CachingPatternSpecFactory.java:72)
        at org.gradle.api.tasks.util.internal.CachingPatternSpecFactory$CachingSpec.isSatisfiedBy(CachingPatternSpecFactory.java:61)
        at org.gradle.api.specs.AndSpec.isSatisfiedBy(AndSpec.java:48)
        at org.gradle.api.internal.file.collections.DefaultDirectoryWalker.shouldVisit(DefaultDirectoryWalker.java:51)
        at org.gradle.api.internal.file.collections.DefaultDirectoryWalker$PathVisitor.preVisitDirectory(DefaultDirectoryWalker.java:88)
        at org.gradle.api.internal.file.collections.DefaultDirectoryWalker$PathVisitor.preVisitDirectory(DefaultDirectoryWalker.java:66)
        at org.gradle.api.internal.file.collections.DefaultDirectoryWalker.walkDir(DefaultDirectoryWalker.java:60)
        at org.gradle.api.internal.file.collections.DirectoryFileTree.walkDir(DirectoryFileTree.java:160)
        at org.gradle.api.internal.file.collections.DirectoryFileTree.visitFrom(DirectoryFileTree.java:138)
        at org.gradle.api.internal.file.collections.DirectoryFileTree.visit(DirectoryFileTree.java:123)
        at org.gradle.api.internal.file.collections.FileTreeAdapter.visit(FileTreeAdapter.java:119)
        at org.gradle.api.internal.file.AbstractFileTree.getFiles(AbstractFileTree.java:58)
        at org.gradle.api.internal.file.CompositeFileCollection.getFiles(CompositeFileCollection.java:76)
        at org.gradle.api.internal.file.CompositeFileCollection.getFiles(CompositeFileCollection.java:54)
        at org.gradle.api.internal.tasks.compile.incremental.recomp.AbstractRecompilationSpecProvider.deleteStaleFilesIn(AbstractRecompilationSpecProvider.java:51)
        at org.gradle.api.internal.tasks.compile.incremental.recomp.JavaRecompilationSpecProvider.initializeCompilation(JavaRecompilationSpecProvider.java:76)
        at org.gradle.api.internal.tasks.compile.incremental.SelectiveCompiler.execute(SelectiveCompiler.java:73)
        at org.gradle.api.internal.tasks.compile.incremental.SelectiveCompiler.execute(SelectiveCompiler.java:36)
        at org.gradle.api.internal.tasks.compile.incremental.IncrementalResultStoringCompiler.execute(IncrementalResultStoringCompiler.java:60)
        at org.gradle.api.internal.tasks.compile.incremental.IncrementalResultStoringCompiler.execute(IncrementalResultStoringCompiler.java:44)
        at org.gradle.api.internal.tasks.compile.CompileJavaBuildOperationReportingCompiler$2.call(CompileJavaBuildOperationReportingCompiler.java:59)
        at org.gradle.api.internal.tasks.compile.CompileJavaBuildOperationReportingCompiler$2.call(CompileJavaBuildOperationReportingCompiler.java:51)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor$CallableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:416)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor$CallableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:406)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor$1.execute(DefaultBuildOperationExecutor.java:165)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:250)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:158)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.call(DefaultBuildOperationExecutor.java:102)
        at org.gradle.internal.operations.DelegatingBuildOperationExecutor.call(DelegatingBuildOperationExecutor.java:36)
        at org.gradle.api.internal.tasks.compile.CompileJavaBuildOperationReportingCompiler.execute(CompileJavaBuildOperationReportingCompiler.java:51)
        at org.gradle.api.tasks.compile.JavaCompile.performCompilation(JavaCompile.java:158)
        at org.gradle.api.tasks.compile.JavaCompile.compile(JavaCompile.java:126)
        at org.gradle.internal.reflect.JavaMethod.invoke(JavaMethod.java:103)
        at org.gradle.api.internal.project.taskfactory.IncrementalTaskInputsTaskAction.doExecute(IncrementalTaskInputsTaskAction.java:47)
        at org.gradle.api.internal.project.taskfactory.StandardTaskAction.execute(StandardTaskAction.java:42)
        at org.gradle.api.internal.project.taskfactory.AbstractIncrementalTaskAction.execute(AbstractIncrementalTaskAction.java:25)
        at org.gradle.api.internal.project.taskfactory.StandardTaskAction.execute(StandardTaskAction.java:28)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter$5.run(ExecuteActionsTaskExecuter.java:476)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:402)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:394)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor$1.execute(DefaultBuildOperationExecutor.java:165)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:250)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:158)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:92)
        at org.gradle.internal.operations.DelegatingBuildOperationExecutor.run(DelegatingBuildOperationExecutor.java:31)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeAction(ExecuteActionsTaskExecuter.java:461)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeActions(ExecuteActionsTaskExecuter.java:444)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.access$200(ExecuteActionsTaskExecuter.java:93)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter$TaskExecution.execute(ExecuteActionsTaskExecuter.java:237)
        at org.gradle.internal.execution.steps.ExecuteStep.lambda$execute$0(ExecuteStep.java:32)
        at org.gradle.internal.execution.steps.ExecuteStep.execute(ExecuteStep.java:32)
        at org.gradle.internal.execution.steps.ExecuteStep.execute(ExecuteStep.java:26)
        at org.gradle.internal.execution.steps.CleanupOutputsStep.execute(CleanupOutputsStep.java:58)
        at org.gradle.internal.execution.steps.CleanupOutputsStep.execute(CleanupOutputsStep.java:35)
        at org.gradle.internal.execution.steps.ResolveInputChangesStep.execute(ResolveInputChangesStep.java:48)
        at org.gradle.internal.execution.steps.ResolveInputChangesStep.execute(ResolveInputChangesStep.java:33)
        at org.gradle.internal.execution.steps.CancelExecutionStep.execute(CancelExecutionStep.java:39)
        at org.gradle.internal.execution.steps.TimeoutStep.executeWithoutTimeout(TimeoutStep.java:73)
        at org.gradle.internal.execution.steps.TimeoutStep.execute(TimeoutStep.java:54)
        at org.gradle.internal.execution.steps.CatchExceptionStep.execute(CatchExceptionStep.java:35)
        at org.gradle.internal.execution.steps.CreateOutputsStep.execute(CreateOutputsStep.java:51)
        at org.gradle.internal.execution.steps.SnapshotOutputsStep.execute(SnapshotOutputsStep.java:45)
        at org.gradle.internal.execution.steps.SnapshotOutputsStep.execute(SnapshotOutputsStep.java:31)
        at org.gradle.internal.execution.steps.CacheStep.executeWithoutCache(CacheStep.java:208)
        at org.gradle.internal.execution.steps.CacheStep.executeAndStoreInCache(CacheStep.java:187)
        at org.gradle.internal.execution.steps.CacheStep.lambda$executeWithCache$2(CacheStep.java:110)
        at org.gradle.internal.execution.steps.CacheStep.lambda$executeWithCache$3(CacheStep.java:110)
        at org.gradle.internal.Try$1.apply(Try.java:59)
        at org.gradle.internal.Try$1.apply(Try.java:56)
        at org.gradle.internal.Try$Success.flatMap(Try.java:102)
        at org.gradle.internal.Try.map(Try.java:56)
        at org.gradle.internal.execution.steps.CacheStep.executeWithCache(CacheStep.java:79)
        at org.gradle.internal.execution.steps.CacheStep.execute(CacheStep.java:69)
        at org.gradle.internal.execution.steps.CacheStep.execute(CacheStep.java:45)
        at org.gradle.internal.execution.steps.BroadcastChangingOutputsStep.execute(BroadcastChangingOutputsStep.java:49)
        at org.gradle.internal.execution.steps.StoreSnapshotsStep.execute(StoreSnapshotsStep.java:43)
        at org.gradle.internal.execution.steps.StoreSnapshotsStep.execute(StoreSnapshotsStep.java:32)
        at org.gradle.internal.execution.steps.RecordOutputsStep.execute(RecordOutputsStep.java:38)
        at org.gradle.internal.execution.steps.RecordOutputsStep.execute(RecordOutputsStep.java:24)
        at org.gradle.internal.execution.steps.SkipUpToDateStep.executeBecause(SkipUpToDateStep.java:96)
        at org.gradle.internal.execution.steps.SkipUpToDateStep.lambda$execute$0(SkipUpToDateStep.java:89)
        at org.gradle.internal.execution.steps.SkipUpToDateStep.execute(SkipUpToDateStep.java:54)
        at org.gradle.internal.execution.steps.SkipUpToDateStep.execute(SkipUpToDateStep.java:38)
        at org.gradle.internal.execution.steps.ResolveChangesStep.execute(ResolveChangesStep.java:76)
        at org.gradle.internal.execution.steps.ResolveChangesStep.execute(ResolveChangesStep.java:37)
        at org.gradle.internal.execution.steps.legacy.MarkSnapshottingInputsFinishedStep.execute(MarkSnapshottingInputsFinishedStep.java:36)
        at org.gradle.internal.execution.steps.legacy.MarkSnapshottingInputsFinishedStep.execute(MarkSnapshottingInputsFinishedStep.java:26)
        at org.gradle.internal.execution.steps.ResolveCachingStateStep.execute(ResolveCachingStateStep.java:90)
        at org.gradle.internal.execution.steps.ResolveCachingStateStep.execute(ResolveCachingStateStep.java:48)
        at org.gradle.internal.execution.steps.CaptureStateBeforeExecutionStep.execute(CaptureStateBeforeExecutionStep.java:69)
        at org.gradle.internal.execution.steps.CaptureStateBeforeExecutionStep.execute(CaptureStateBeforeExecutionStep.java:47)
        at org.gradle.internal.execution.impl.DefaultWorkExecutor.execute(DefaultWorkExecutor.java:33)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.execute(ExecuteActionsTaskExecuter.java:140)
        ... 35 more
Caused by: java.lang.StackOverflowError
        at org.gradle.api.internal.file.pattern.PatternMatcher$3.test(PatternMatcher.java:52)
        at org.gradle.api.internal.file.pattern.PatternMatcher$3.test(PatternMatcher.java:52)
        at org.gradle.api.internal.file.pattern.PatternMatcher$3.test(PatternMatcher.java:52)
...
```

### Context
The StackOverflowException either breaks the build w/ the annoation-processor https://github.com/snazy/compile-command-annotations/tree/fix-gradle-incr-compilation adopted for incremental builds.

### Steps to Reproduce 
Let me know, if you need a reproducer.

--- a/subprojects/files/src/main/java/org/gradle/api/internal/file/pattern/PatternMatcher.java
+++ b/subprojects/files/src/main/java/org/gradle/api/internal/file/pattern/PatternMatcher.java
@@ -16,6 +16,9 @@
 
 package org.gradle.api.internal.file.pattern;
 
+import java.util.ArrayList;
+import java.util.List;
+
 public abstract class PatternMatcher {
     public static final PatternMatcher MATCH_ALL = new PatternMatcher() {
         @Override
@@ -37,21 +40,11 @@ public abstract class PatternMatcher {
     public abstract boolean test(String[] segments, boolean isFile);
 
     public PatternMatcher and(final PatternMatcher other) {
-        return new PatternMatcher() {
-            @Override
-            public boolean test(String[] segments, boolean isFile) {
-                return PatternMatcher.this.test(segments, isFile) && other.test(segments, isFile);
-            }
-        };
+        return new And(PatternMatcher.this, other);
     }
 
     public PatternMatcher or(final PatternMatcher other) {
-        return new PatternMatcher() {
-            @Override
-            public boolean test(String[] segments, boolean isFile) {
-                return PatternMatcher.this.test(segments, isFile) || other.test(segments, isFile);
-            }
-        };
+        return new Or(PatternMatcher.this, other);
     }
 
     public PatternMatcher negate() {
@@ -61,5 +54,55 @@ public abstract class PatternMatcher {
                 return !PatternMatcher.this.test(segments, isFile);
             }
         };
+    }
+
+    private static final class Or extends PatternMatcher {
+        private final List<PatternMatcher> parts = new ArrayList<PatternMatcher>();
+
+        public Or(PatternMatcher patternMatcher, PatternMatcher other) {
+            parts.add(patternMatcher);
+            parts.add(other);
+        }
+
+        @Override
+        public PatternMatcher or(PatternMatcher other) {
+            parts.add(other);
+            return this;
+        }
+
+        @Override
+        public boolean test(String[] segments, boolean isFile) {
+            for (PatternMatcher part : parts) {
+                if (part.test(segments, isFile)) {
+                    return true;
+                }
+            }
+            return false;
+        }
+    }
+
+    private static final class And extends PatternMatcher {
+        private final List<PatternMatcher> parts = new ArrayList<PatternMatcher>();
+
+        public And(PatternMatcher patternMatcher, PatternMatcher other) {
+            parts.add(patternMatcher);
+            parts.add(other);
+        }
+
+        @Override
+        public PatternMatcher and(PatternMatcher other) {
+            parts.add(other);
+            return this;
+        }
+
+        @Override
+        public boolean test(String[] segments, boolean isFile) {
+            for (PatternMatcher part : parts) {
+                if (!part.test(segments, isFile)) {
+                    return false;
+                }
+            }
+            return true;
+        }
     }
 }

--- a/subprojects/files/src/test/groovy/org/gradle/api/internal/file/pattern/PatternMatcherFactoryTest.java
+++ b/subprojects/files/src/test/groovy/org/gradle/api/internal/file/pattern/PatternMatcherFactoryTest.java
@@ -22,12 +22,51 @@ import org.hamcrest.Description;
 import org.hamcrest.Matcher;
 import org.junit.Test;
 
+import java.util.LinkedHashSet;
+import java.util.Set;
+
 import static org.hamcrest.CoreMatchers.instanceOf;
 import static org.hamcrest.CoreMatchers.not;
 import static org.junit.Assert.assertThat;
 
 public class PatternMatcherFactoryTest {
     private PatternMatcher matcher;
+
+    @Test public void testNoStackOverflowForManyPatterns() {
+        // The only reason for this unit test is to verify that no StackOverflowException is being thrown when
+        // many patterns are passed to getPatternsMatcher. See https://github.com/gradle/gradle/issues/10329
+        Set<String> manyPatterns =  new LinkedHashSet<String>();
+        for (int i = 0; i < 5000; i++) {
+            manyPatterns.add("some/package/Some" + i + "ClassName.class");
+            manyPatterns.add("some/package/Some" + i + "ClassName.java");
+            manyPatterns.add("some/package/Some" + i + "ClassName.h");
+            manyPatterns.add("some/package/Some" + i + "ClassName$*.class");
+            manyPatterns.add("some/package/Some" + i + "ClassName$*.java");
+            manyPatterns.add("some/package/Some" + i + "ClassName$*.h");
+        }
+        matcher = PatternMatcherFactory.getPatternsMatcher(true, true, manyPatterns);
+        assertThat(matcher, not(matchesFile("some", "package", "SomeClassName.java")));
+        assertThat(matcher, matchesFile("some", "package", "Some123ClassName.java"));
+    }
+
+    @Test public void testManyOr() {
+        Set<String> manyPatterns =  new LinkedHashSet<String>();
+        manyPatterns.add("some/package/SomeClassName.class");
+        manyPatterns.add("some/package/SomeClassName.java");
+        manyPatterns.add("some/package/SomeClassName.h");
+        manyPatterns.add("some/package/SomeClassName$*.class");
+        manyPatterns.add("some/package/SomeClassName$*.java");
+        manyPatterns.add("some/package/SomeClassName$*.h");
+        matcher = PatternMatcherFactory.getPatternsMatcher(true, true, manyPatterns);
+        assertThat(matcher, not(matchesFile()));
+        assertThat(matcher, matchesFile("some", "package", "SomeClassName.java"));
+        assertThat(matcher, matchesFile("some", "package", "SomeClassName.class"));
+        assertThat(matcher, matchesFile("some", "package", "SomeClassName.h"));
+        assertThat(matcher, matchesFile("some", "package", "SomeClassName$*.java"));
+        assertThat(matcher, matchesFile("some", "package", "SomeClassName$*.class"));
+        assertThat(matcher, matchesFile("some", "package", "SomeClassName$*.h"));
+        assertThat(matcher, not(matchesFile("a")));
+    }
 
     @Test public void testEmpty() {
         matcher = PatternMatcherFactory.getPatternMatcher(true, true, "");

